src/main/java/org/prebid/server/bidder/huaweiads/HuaweiAdSlotBuilder.java

```diff
@@ -141,13 +147,80 @@ private List<Asset> parseNativeRequestAssets(Native xNative) {
     private static List<String> makeDetailedCreativeTypeList(long numVideo, long numImage) {
         if (numVideo >= 1) {
             return List.of("903");
-        } else if (numImage > 1) {
-            return List.of("904");
-        } else if (numImage == 1) {
-            return List.of("901");
+        } else if (numImage >= 1) {
+            return List.of("901", "904", "905");
         } else {
             return List.of("913", "914");
         }
     }
 
+    private List<Format> makeFormatListForNative(List<Asset> assets, long numImage) {
+        final Set<Format> formats = new HashSet<>();
+        formats.addAll(makeFormatListForVideo(assets));
+        formats.addAll(makeFormatListForImage(assets, numImage));
+        return formats.stream().toList();
+    }
+
+    private List<Format> makeFormatListForVideo(List<Asset> assets) {
+        return assets.stream()
+                .map(Asset::getVideo)
+                .filter(Objects::nonNull)
+                .filter(video -> HuaweiUtils.isFormatDefined(video.getW(), video.getH()))
+                .map(video -> Format.of(video.getW(), video.getH()))
+                .distinct()
+                .toList();
+    }
+
+    private Set<Format> makeFormatListForImage(List<Asset> assets, long numImage) {
+        final Set<Format> formats = new HashSet<>();
+        assets.stream()
+                .map(Asset::getImg)
+                .filter(Objects::nonNull)
+                .filter(image -> HuaweiUtils.isFormatDefined(image.getW(), image.getW()))
+                .forEach(image -> {
+                    if (numImage > 1
+                            && HuaweiUtils.isFormatDefined(image.getW(), image.getH())
+                            && HuaweiUtils.isFormatDefined(image.getWmin(), image.getHmin())) {
+                        formats.add(Format.of(image.getW(), image.getH()));
+                    }
+                    if (numImage == 1
+                            && HuaweiUtils.isFormatDefined(image.getW(), image.getH())
+                            && HuaweiUtils.isFormatDefined(image.getWmin(), image.getHmin())) {
+                        formats.addAll(filterPopularSizes(image.getW(), image.getH(), true));
+                    }
+                    if (numImage == 1
+                            && !HuaweiUtils.isFormatDefined(image.getW(), image.getH())
+                            && HuaweiUtils.isFormatDefined(image.getWmin(), image.getHmin())) {
+                        formats.addAll(filterPopularSizes(image.getWmin(), image.getHmin(), false));
+                    }
+                });
+        return formats;
+    }
+
+    private static List<Format> filterPopularSizes(Integer width, Integer height, boolean filterByRatio) {
+        if (filterByRatio) {
+            return getPopularFormats().stream()
+                    .filter(size ->
+                            Math.abs((double) width / height - (double) size.getW() / size.getH()) <= FILL_RATIO)
```

Comment:
working with doubles might cause problems, I would suggest re-write it to use BigDecimals

Reply:
Fixed.